modules/502-delivery/hooks/werf_sources/hook.go

```diff
@@ -0,0 +1,406 @@
+
+		argoRepos = append(argoRepos, argocdHelmOCIRepository{
+			Name:     ws.name,
+			Username: username,
+			Password: password,
+			Project:  ws.argocdRepo.project,
+			URL:      ws.repo,
+		})
+	}
+	return argoRepos, nil
+}
+
+func extractCredentials(credentialsBySecretName map[string]dockerFileConfig, pullSecretName, registry string) (string, string) {
+	username, password := "", ""
+	config, ok := credentialsBySecretName[pullSecretName]
+	if !ok {
+		return username, password
+	}
+
+	creds, err := parseDockerConfigJSONCredentials(config, registry)
+	if err != nil {
+		return username, password
+	}
+	return creds.username, creds.password
+}
+
+// cr.example.com/path/to/image -> cr.example.com
+func firstSegment(s string) string {
+	for i, c := range s {
+		if c == '/' {
+			return s[:i]
+		}
+	}
+	return s
+}
+
+func filterWerfSource(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var (
+		ws  werfSource
+		err error
+		ok  bool
+	)
+
+	ws.name = obj.GetName()
+
+	ws.repo, ok, err = unstructured.NestedString(obj.Object, "spec", "imageRepo")
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		return nil, fmt.Errorf("spec.imageRepo field expected")
+	}
+
+	ws.apiURL, ok, err = unstructured.NestedString(obj.Object, "spec", "apiUrl")
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		ws.apiURL = "https://" + firstSegment(ws.repo)
+	}
+
+	ws.pullSecretName, _, err = unstructured.NestedString(obj.Object, "spec", "pullSecretName")
+	if err != nil {
+		return nil, err
+	}
+
+	// By default, Argo CD is desired, but the OCI repo can be disabled to use purely Image
+	// Updater functionality along with another repository type.
+	repoEnabled, ok, err := unstructured.NestedBool(obj.Object, "spec", "argocdRepoEnabled")
+	if err != nil {
+		return nil, err
+	}
+	if !ok {
+		repoEnabled = true
+	}
+
+	// By default, Argo CD repo belongs to the "default" project.
+	arepo, ok, err := unstructured.NestedStringMap(obj.Object, "spec", "argocdRepo")
+	if err != nil {
+		return nil, err
+	}
+	project := "default"
+	if repoEnabled && ok {
+		specifiedProject, projectSpecified := arepo["project"]
+		if projectSpecified && specifiedProject != "" {
+			project = specifiedProject
+		}
+	}
+	if repoEnabled {
+		ws.argocdRepo = &argocdRepoConfig{
+			project: project,
+		}
+	}
+
+	return ws, nil
+}
+
+func parseWerfSources(snapshots []go_hook.FilterResult) ([]werfSource, error) {
+	var res []werfSource
+	for _, snap := range snapshots {
+		r, ok := snap.(werfSource)
+		if !ok {
+			return nil, fmt.Errorf("unexpected type %T", snap)
+		}
+		res = append(res, r)
+	}
+	return res, nil
+}
+
+type registryCredentials struct {
+	username string
+	password string
+}
+
+func parseDockerConfigJSONCredentials(config dockerFileConfig, registry string) (registryCredentials, error) {
+	creds := registryCredentials{}
+
+	cfg, ok := config.Auths[registry]
+	if !ok {
+		return creds, fmt.Errorf("no credentials")
+	}
+
+	if cfg.Auth != "" {
```

Comment:
Should Deckhouse fall back to username/password if Auth is in the wrong format instead of throwing an error?

Reply:
Because the delivery module uses secrets with the type that was originally used by kubelet to pull images, I'd like to suggest keeping the same logic as in kubelet.